piquasso/api/instruction.py

```diff
@@ -204,6 +221,49 @@ def _validate_modes(self, modes):
                 f"{self.NUMBER_OF_MODES}'."
             )
 
+    def when(self, condition: Callable) -> "Instruction":
+        """Adds a callable condition to the instruction.
+
+        The callable takes a single argument, which is the list of measurement outcomes,
+        and returns a boolean indicating whether the condition is met. All conditions
+        must be met for the instruction to be executed.
+
+        The measurement outcomes are provided as a tuple, in the order which the
+        measurements are defined. For example, `x[0]` refers to the first outcome, and
+        `x[-1]` to the last outcome.
+
+        Example usage::
+            with pq.Program() as program:
+                pq.Q(0, 1) | pq.StateVector([0, 2]) * np.sqrt(1 / 2)
+                pq.Q(0, 1) | pq.StateVector([2, 0]) * np.sqrt(1 / 2)
+
+                pq.Q(1)    | pq.ParticleNumberMeasurement().on_modes(1)
+                pq.Q(0)    | pq.Squeezing(r=r).on_modes(0).when(lambda x: x[-1] == 2)
+
+                pq.Q(0, 1) | pq.Beamsplitter5050().when(lambda x: x[0] > 0)
+
+        Args:
+            condition (Callable):
+                A callable that takes the measurement outcomes and returns a boolean.
+        """
+        if not callable(condition):
+            raise PiquassoException(
+                f"The condition '{condition}' needs to be a callable that takes the "
+                f"measurement outcomes and returns a boolean."
+            )
+
+        self._conditions.append(condition)
+
+        return self
+
+    def _are_conditions_met(self, outcome: Tuple[Union[int, float], ...]) -> bool:
```

Comment:
We had discussions about terminologies, but again, in my book, "outcome" is a singular object i.e., a number and a tuple is unexpected for me here.

Reply:
What name do you suggest here?